### PR TITLE
feat: ESLint v9 Support

### LINKED ISSUE #7286: eslint-plugin-query is incompatible with Eslint 9.x
### Describe the bug

eslint-plugin-query has a peer dependency for eslint 8.
When I try to upgrade my project to use eslint 9, npm install fails due to peer dependency mismatch.

Thanks!

### Your minimal, reproducible example

Any NodeJS project

### Steps to reproduce

1. `npm i -D @tanstack/eslint-plugin-query@5.28.11`
2. `npm i -D eslint@9.0.0`
Erroneous output:
```
npm ERR! While resolving: eslint-import-resolver-typescript@3.6.1
npm ERR! Found: eslint@9.0.0
npm ERR! node_modules/eslint
npm ERR!   peer eslint@"^7.0.0 || ^8.0.0" from @typescript-eslint/utils@6.21.0
npm ERR!   node_modules/@tanstack/eslint-plugin-query/node_modules/@typescript-eslint/utils
npm ERR!     @typescript-eslint/utils@"^6.20.0" from @tanstack/eslint-plugin-query@5.28.11
npm ERR!     node_modules/@tanstack/eslint-plugin-query
npm ERR!       dev @tanstack/eslint-plugin-query@"^5.28.11" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer eslint-plugin-import@"*" from eslint-import-resolver-typescript@3.6.1
npm ERR! node_modules/eslint-import-resolver-typescript
npm ERR!   dev eslint-import-resolver-typescript@"^3.6.1" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: eslint@8.57.0
npm ERR! node_modules/eslint
npm ERR!   peer eslint@"^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8" from eslint-plugin-import@2.29.1
npm ERR!   node_modules/eslint-plugin-import
npm ERR!     peer eslint-plugin-import@"*" from eslint-import-resolver-typescript@3.6.1
npm ERR!     node_modules/eslint-import-resolver-typescript
npm ERR!       dev eslint-import-resolver-typescript@"^3.6.1" from the root project
```

### Expected behavior

Installation should be smooth :)

### How often does this bug happen?

None

### Screenshots or Videos

_No response_

### Platform

all

### Tanstack Query adapter

None

### TanStack Query version

5.28.11

### TypeScript version

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by davidjbng**
Closes #7286

This PR adds support for ESLint v9 and is backwards compatible with ESLint v8.

I upgraded to the @typescript-eslint v8 alpha packages which support ESLint v9, but that required some type changes.

It is mostly the same as proposed in #7253 but **does not yet support the flat config format**. I will create a separate tracking issue for that.

## Tests

- [x] Tested with ESLint v8 - the changes are backwards compatible (afaik)
- [x] Tested with ESLint v9 - https://github.com/davidjbng/eslint-plugin-query-v9-example

## To Be Discussed 
- [ ] Are the @typescript-eslint v8 alpha packages ok to merge or should we wait for a stable release?

## Docs

I did not update any docs yet because there are basically no configuration changes required to use v9. 
Flat config will be more important to document once it is supported.

### CODE REVIEW

File: packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts [hunk 1]
```diff
@@ -679,33 +679,6 @@ ruleTester.run('exhaustive-deps', rule, {
         },
       ],
     },
-    {
-      name: 'should fail when a queryKey is a reference of an array expression with a missing dep',
```
Comment (TkDodo) [inline packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts:683]: why was this test removed?

**Review by TkDodo — COMMENTED** [review]
Thank you 🙏 . Is this fully backwards compatible? Do we need to change the docs for users using the flat / legacy configs?

**Comment by TkDodo** [comment]
please also update eslint in the examples. It would serve as a good reference point to see if things are still working

File: packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts (hunk 1 above)
Comment (davidjbng) [reply packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts:683]: typescript-eslint now detects duplicated test cases (https://github.com/typescript-eslint/typescript-eslint/issues/9018) so this was failing after upgrading.

I don't know why it was duplicated in the first place.

https://github.com/TanStack/query/blob/f353ef8d3d3973ac6de8155eb69cb943dcdbb8df/packages/eslint-plugin-query/src/__tests__/exhaustive-deps.test.ts#L655-L708

**Comment by davidjbng** [comment]
> Thank you 🙏 . Is this fully backwards compatible? Do we need to change the docs for users using the flat / legacy configs?

I am not aware flat config could be used before.

We could update the docs to say that ESLint v9 is supported with `ESLINT_USE_FLAT_CONFIG=false` if that is desired, but I would expect flat config support is added soon where I would add a full explanation on the new format.

In my test repo this was a non-breaking update with the following setup:
Legacy cjs config (`.eslintrc.cjs`) with the following package versions.
```json
    "@typescript-eslint/eslint-plugin": "7.8.0",
    "@typescript-eslint/parser": "7.8.0",
    "eslint": "8.57.0",
    "eslint-import-resolver-typescript": "3.6.1",
    "eslint-plugin-import": "2.29.1",
    "eslint-plugin-jest": "27.9.0",
    "eslint-plugin-react": "7.34.1",
    "eslint-plugin-react-hooks": "4.6.2",
```

I installed the updated pluging from file after running `pnpm build` on the changes in this pr ("@tanstack/eslint-plugin-query": "file:../../query/packages/eslint-plugin-query").

**Comment by TkDodo** [comment]
okay thanks, so this is just for v9 support, _not_ for flat config support?

**Comment by davidjbng** [comment]
> please also update eslint in the examples. It would serve as a good reference point to see if things are still working

I tried to create a new example at first, but I was unsure on how to reference the plugin updates I had to do because the other examples were referencing published package versions.

I created this repo https://github.com/davidjbng/eslint-plugin-query-v9-example which is just a copy of basic-typescript with eslint v9 config

**Comment by davidjbng** [comment]
> okay thanks, so this is just for v9 support, _not_ for flat config support?

Yes. I opened a new issue for Flat Config here: https://github.com/TanStack/query/issues/7544
It should not be a big change, but I wanted to keep things minimal to support v9 first.

**Review by TkDodo — APPROVED** [review]

**Comment by TkDodo** [comment]
okay the error that is left is from `sherif`, the tool we use to streamline dependencies in the monorepo. As discussed with other maintainers, I think the best move is to completely exclude `examples` and `integrations` from the `sherif` check.

I did that here: a42a2ed2

**Comment by TkDodo** [comment]
alright, let' ship it

**Comment by MWhite-22** [comment]
FYI getting the below peer-dependency warning with this release. @TkDodo @davidjbng 

image

I think the peer definition should be `>=8` but I am not 100% sure.

**Comment by TkDodo** [comment]
yes, someone please fix it. It should be:

```
"eslint": "^8 || ^9"
```

just like we have for react:

```
"react": "^18 || ^19"
```

**Comment by davidjbng** [comment]
Working on the fix.

Seems like npm fails to report peerDependency errors when installing from a file path. I found at least a few similar issues like this one: https://github.com/npm/cli/issues/7143.

**Pull request merged**